src/utils/SparseGrid.h

```diff
@@ -0,0 +1,360 @@
+     * an area of (2*radius + cell_size)**2 on average.  The max range
+     * of an item from the query_point is radius + cell_size.
+     *
+     * \param[in] query_pt The point to search around.
+     * \param[in] radius The search radius.
+     * \return Vector of elements found
+     */
+    std::vector<Elem> getNearby(const Point &query_pt, coord_t radius) const;
+
+    static const std::function<bool(const Elem&)> no_precondition;
+
+    /*!
+     * Find the nearest element to a given \p query_pt within \p radius.
+     *
+     * \param[in] query_pt The point for which to find the nearest object.
+     * \param[in] radius The search radius.
+     * \param[out] elem_nearest the nearest element. Only valid if function returns true.
+     * \param[in] precondition A precondition which must return true for an element
+     *    to be considered for output
+     * \return True if and only if an object has been found within the radius.
+     */
+    bool getNearest(const Point &query_pt, coord_t radius, Elem &elem_nearest,
+                    const std::function<bool(const Elem& elem)> precondition = no_precondition) const;
+
+    /*! \brief Process elements from cells that might contain sought after points.
+     *
+     * Processes elements from cell that might have elements within \p
+     * radius of \p query_pt.  Processes all elements that are within
+     * radius of query_pt.  May process elements that are up to radius +
+     * cell_size from query_pt.
+     *
+     * \param[in] query_pt The point to search around.
+     * \param[in] radius The search radius.
+     * \param[in] process_func Processes each element.  process_func(elem) is
+     *    called for each element in the cell.
+     */
+    template<class ProcessFunc>
+    void processNearby(const Point &query_pt, coord_t radius,
+                       ProcessFunc &process_func) const;
+
+    coord_t getCellSize() const;
+
+private:
+    using GridPoint = Point;
+    using GridMap = std::unordered_multimap<GridPoint, Elem>;
+
+    /*! \brief Process elements from the cell indicated by \p grid_pt.
+     *
+     * \param[in] grid_pt The grid coordinates of the cell.
+     * \param[in] process_func Processes each element.  process_func(elem) is
+     *    called for each element in the cell.
+     */
+    template<class ProcessFunc>
+    void processFromCell(const GridPoint &grid_pt,
+                         ProcessFunc &process_func) const;
+
+    /*! \brief Compute the grid coordinates of a point.
+     *
+     * \param[in] point The actual location.
+     * \return The grid coordinates that correspond to point.
+     */
+    GridPoint toGridPoint(const Point &point) const;
+
+    /*! \brief Map from grid locations (GridPoint) to elements (Elem). */
+    GridMap m_grid;
+    /*! \brief Accessor for getting locations from elements. */
+    PointAccess m_point_access;
+    /*! \brief The cell (square) size. */
+    coord_t m_cell_size;
+};
+
+namespace SparseGridImpl {
```

Comment:
I wanted to make it private, but didn't have a good way to do that.  Putting it in a namespace gave me a way to at least discourage use from outside the module.